// react-native-embrace/dist/scripts/util/wizard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Wizard {
    constructor() {
        this.fields = {};
        this.steps = [];
        this.fieldValues = {};
    }
    registerField(field) {
        this.fields[field.name] = field;
    }
    registerStep(step) {
        this.steps = this.steps.concat(step);
    }
    fieldValue(field) {
        const value = this.fieldValues[field.name];
        if (value !== undefined) {
            return Promise.resolve(value);
        }
        return field.fetch().then((val) => {
            this.fieldValues[field.name] = val;
            return val;
        });
    }
    fieldValueList(list) {
        return list.reduce((chain, field) => chain.then((results) => field.fetch().then((res) => [...results, res])), Promise.resolve([]));
    }
    runSteps() {
        return this.steps.reduce((chain, step) => chain.then((results) => step.run(this).then((res) => [...results, res])), Promise.resolve([]));
    }
}
exports.default = Wizard;
